org.eclipse.xtext/src/org/eclipse/xtext/resource/impl/ChunkedResourceDescriptions.java

```diff
@@ -0,0 +1,207 @@
+/**
+ * Copyright (c) 2015, 2020 itemis AG (http://www.itemis.eu) and others.
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0.
+ * 
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.eclipse.xtext.resource.impl;
+
+import static com.google.common.collect.Iterables.*;
+import static org.eclipse.xtext.xbase.lib.IterableExtensions.*;
+
+import java.io.Externalizable;
+import java.io.IOException;
+import java.io.ObjectInput;
+import java.io.ObjectOutput;
+import java.io.Serializable;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.concurrent.ConcurrentHashMap;
+
+import org.eclipse.emf.common.notify.Adapter;
+import org.eclipse.emf.common.notify.Notifier;
+import org.eclipse.emf.common.notify.impl.AdapterImpl;
+import org.eclipse.emf.common.util.URI;
+import org.eclipse.emf.ecore.resource.ResourceSet;
+import org.eclipse.xtext.resource.IResourceDescription;
+import org.eclipse.xtext.resource.IResourceDescriptions;
+import org.eclipse.xtext.resource.ISelectable;
+import org.eclipse.xtext.resource.containers.ProjectDescriptionBasedContainerManager;
+import org.eclipse.xtext.resource.persistence.SerializableResourceDescription;
+import org.eclipse.xtext.xbase.lib.ExclusiveRange;
+import org.eclipse.xtext.xbase.lib.IterableExtensions;
+
+import com.google.common.annotations.Beta;
+
+/**
+ * A IResourceDescriptions implementation that holds its resource description in chunks, each identified by a string.
+ * The strings represent units such as projects, source sets, and libraries.
+ * 
+ * @see ProjectDescription
+ * @see ProjectDescriptionBasedContainerManager
+ * 
+ * @author Sven Efftinge - Initial contribution and API
+ * 
+ * @since 2.9
+ */
+@Beta
+public class ChunkedResourceDescriptions extends AbstractCompoundSelectable
+		implements IResourceDescriptions, Externalizable {
+
+	protected ConcurrentHashMap<String, ResourceDescriptionsData> chunk2resourceDescriptions = new ConcurrentHashMap<>();
+	protected ResourceSet resourceSet;
+
+	public ChunkedResourceDescriptions() {
+	}
+
+	public ChunkedResourceDescriptions(Map<String, ResourceDescriptionsData> initialData) {
+		chunk2resourceDescriptions = new ConcurrentHashMap<>(initialData);
+	}
+
+	public ChunkedResourceDescriptions(Map<String, ResourceDescriptionsData> initialData, ResourceSet resourceSet) {
+		this(initialData);
+		setResourceSet(resourceSet);
+	}
+
+	/**
+	 * Creates a shallow copy of the resource descriptions map and installs it with the given ResourceSet.
+	 */
+	public ChunkedResourceDescriptions createShallowCopyWith(ResourceSet resourceSet) {
+		return new ChunkedResourceDescriptions(chunk2resourceDescriptions, resourceSet);
+	}
+
+	public ResourceSet getResourceSet() {
+		return this.resourceSet;
```

Comment:
Few unneeded `this` in this file

Reply:
Only one ... the others are needed to avoid name clashes.